### PR TITLE
Add delete/select word before/after

### PR DESCRIPTION
**Pull request by andrechen77**
Add keybindings for delete word before/after and select word before/after.

Mimics the behavior of Ctrl+Backspace/Delete and Ctrl+Shift+Left/Right on Windows, and the same but on Alt for Mac.

### CODE REVIEW

File: src/actions.ts [hunk 1]
```diff
@@ -511,6 +513,40 @@ export const moveWord = (editor: Editor, direction: 'left' | 'right') => {
...
+    .map((selection) => selection.head);
+  moveWord(editor, direction === 'before' ? 'left' : 'right');
+  const newPositions = editor
+    .listSelections()
+    .map((selection) => selection.head);
+
+  const changes: EditorChange[] = oldPositions.map((oldPos, index) => {
+    const newPos = newPositions[index];
+    const [from, to] =
+      direction === 'before' ? [newPos, oldPos] : [oldPos, newPos];
```
Comment (timhor) [inline src/actions.ts:532]: This would misalign if multiple cursors merge. Example: trigger 'delete word before cursor' on the word `testing` with a cursor halfway through and another cursor at the end. The delete doesn't operate as expected.

File: src/actions.ts [hunk 2]
```diff
@@ -511,6 +513,40 @@ export const moveWord = (editor: Editor, direction: 'left' | 'right') => {
...
+    return { from, to, text: '' };
+  });
+  editor.transaction({ changes });
+};
+
+// There is no built-in command for extending the selection by one word, so
+// the anchor of each selection is preserved across a plain word-move
+export const selectWord = (editor: Editor, direction: 'left' | 'right') => {
```
Comment (timhor) [inline src/actions.ts:540]: This is ambiguously named with `selectWordOrNextOccurrence`. Could I suggest `selectToWordInDirection`? And perhaps update the delete action to a corresponding `deleteWordInDirection`?

Comment (timhor) [inline src/actions.ts:540]: Is this supposed to stop extending after the first press? It still remains on the same selected word even after repeated invocations

File: src/actions.ts [hunk 3]
```diff
@@ -511,6 +513,40 @@ export const moveWord = (editor: Editor, direction: 'left' | 'right') => {
   }
 };
 
+// There is no built-in command for deleting a word (unlike goWordLeft /
+// goWordRight, delWordBefore / delWordAfter are not supported by Obsidian's
+// editor.exec), so the word boundary is instead found by moving the cursor
+// with moveWord and deleting the text passed over
+export const deleteWord = (editor: Editor, direction: 'before' | 'after') => {
```
Comment (timhor) [inline src/actions.ts:520]: Should this still work with a non-empty selection? Currently nothing happens. Example: trigger 'delete word after cursor' with several characters highlighted on a word.

File: src/actions.ts [hunk 4]
```diff
@@ -628,6 +664,150 @@ export const expandSelectionToQuotesOrBrackets = (editor: Editor) => {
...
+        candidateFrom <= selectionRange.from &&
+        candidateTo >= selectionRange.to &&
+        (candidateFrom < selectionRange.from ||
+          candidateTo > selectionRange.to),
+    )
+    .sort((a, b) => a.to - a.from - (b.to - b.from))[0];
+
+  if (nextRange) {
+    editor.setSelection(
+      editor.offsetToPos(nextRange.from),
```
Comment (timhor) [inline src/actions.ts:805]: `metadataCache` may be stale while it debounces, should we clamp this to a max value of `documentEndOffset`?

File: src/actions.ts [hunk 5]
```diff
@@ -628,6 +664,150 @@ export const expandSelectionToQuotesOrBrackets = (editor: Editor) => {
...
+ * Expands the selection to the nearest enclosing syntactic scope, similar to
+ * VS Code's 'Expand Selection' command. Rather than walking a parse tree
+ * (which Obsidian doesn't expose publicly), this gathers every range known
+ * from the document's cached metadata plus a bracket/quote character scan,
+ * and jumps to whichever of those ranges is the smallest one that still
+ * properly contains the current selection. This naturally reconstructs
+ * nesting (e.g. list item -> parent list item -> enclosing blockquote ->
+ * heading section) without needing to know the 'correct' order up front.
+ */
+export const expandSelectionToScope = (app: App, editor: Editor) => {
```
Comment (timhor) [inline src/actions.ts:740]: A few things here:
- Are we missing a word-level step? It currently goes from cursor to the entire line immediately, I'd expect it to select the current word first
- Should it be possible to include bracket pairs in the selection? Running this on `(test)` only selects the inner `test` and there is no step where it includes the parentheses.
- This discards other cursors in a multi-cursor selection. Can we make it support the same action for each active selection?

File: src/actions.ts [hunk 6]
```diff
@@ -628,6 +664,150 @@ export const expandSelectionToQuotesOrBrackets = (editor: Editor) => {
...
+
+  const nextRange = candidates
+    .filter(
+      ({ from: candidateFrom, to: candidateTo }) =>
+        candidateFrom <= selectionRange.from &&
+        candidateTo >= selectionRange.to &&
+        (candidateFrom < selectionRange.from ||
+          candidateTo > selectionRange.to),
+    )
+    .sort((a, b) => a.to - a.from - (b.to - b.from))[0];
```
Comment (timhor) [inline src/actions.ts:801]: Could we replace this with a single `reduce` call? Saves on memory allocation + sort operation just to take one element. Something like:

```ts
const width = ({ from, to }: OffsetRange) => to - from;

const nextRange = candidates.reduce<OffsetRange | null>((smallest, candidate) => {
  const contains =
    candidate.from <= selectionRange.from &&
    candidate.to >= selectionRange.to &&
    (candidate.from < selectionRange.from || candidate.to > selectionRange.to);
  if (!contains) {
    return smallest;
  }
  return !smallest || width(candidate) < width(smallest) ? candidate : smallest;
}, null);
```

File: src/actions.ts [hunk 7]
```diff
@@ -628,6 +664,150 @@ export const expandSelectionToQuotesOrBrackets = (editor: Editor) => {
...
 
+type OffsetRange = { from: number; to: number };
+
+/**
+ * Computes the full span of the section introduced by a heading: from the
+ * start of the heading line up to (but excluding) the next heading of an
+ * equal or shallower level, or the end of the document if there is none.
+ */
+const getHeadingSectionRanges = (
+  editor: Editor,
```
Comment (timhor) [inline src/actions.ts:675]: We don't need the `editor` param here right?

File: src/__tests__/actions-cursor.spec.ts
Comment (timhor) [inline src/__tests__/actions-cursor.spec.ts:1]: Can we add some scenarios in `actions-range` and `actions-multi` as well? It may surface some missing behavioural gaps (as called out below) 🙏

File: src/__tests__/actions-cursor.spec.ts [hunk 8]
```diff
@@ -468,6 +483,69 @@ describe('Code Editor Shortcuts: actions - single cursor selection', () => {
...
+      editor.setCursor({ line: 1, ch: 0 }); // '{cursor}dolor sit'
+      deleteWord(editor as any, 'after');
+
+      const { doc, cursor } = getDocumentAndSelection(editor);
+      expect(doc).toEqual('lorem ipsum\n sit\namet');
+      expect(cursor).toEqual(expect.objectContaining({ line: 1, ch: 0 }));
+    });
+  });
+
+  describe('selectWord', () => {
```
Comment (timhor) [inline src/__tests__/actions-cursor.spec.ts:519]: `editor.exec` is still a no-op mock here. We'll need to add in the same forwarding logic as `deleteWord`'s `beforeEach` to make this a useful test, otherwise it's not really asserting on anything

**Review by timhor — CHANGES_REQUESTED** [review]
Thanks for the contribution 😀 I noticed a few edge cases and areas to address, if you want to take another look

File: src/actions.ts (hunk 2 above)
Comment (andrechen77) [reply src/actions.ts]: Fixed.

File: src/actions.ts (hunk 1 above)
Comment (andrechen77) [reply src/actions.ts]: Fixed.

File: src/actions.ts (hunk 2 above)
Comment (andrechen77) [reply src/actions.ts]: Yessir

File: src/actions.ts (hunk 3 above)
Comment (andrechen77) [reply src/actions.ts]: Fixed.

File: src/__tests__/actions-cursor.spec.ts
Comment (andrechen77) [reply src/__tests__/actions-cursor.spec.ts:1]: Done.

File: src/__tests__/actions-cursor.spec.ts (hunk 8 above)
Comment (andrechen77) [reply src/__tests__/actions-cursor.spec.ts]: Fixed.

**Comment by andrechen77** [comment]
Thanks for the review! The issues that you pointed out should be addressed now.

The commit with the feature of "expanding scope" was not intended to be included; that was a personal experiment. I've removed it from the PR with this update, but I can bring it back with fixes if you'd like.
